### ISSUE TITLE
Feature Requests:  Get Embedded Metadata from Models, and Allow Customizing Metadata in Image Drawer

### ISSUE THREAD

**Issue by Monty1836** [JaredTherriault/ComfyUI-JNodes#19]
Safetensors models generally include metadata containing information about the model in the header.  This includes useful things such as a name, training image resolutions, and tags used for training.  It would be very useful to have a button in the dropdown for models to parse, and display the embedded metadata in a friendly way.   It would also be good to have an option to display the embedded model name as the model's friendly name in the image drawer rather than only one specified in an external metadata file.

I don't use a downloader to install my models, so I don't normally have metadata files downloaded alongside them.  That means all models just use the name of the safetensors file, rather than the actual name the model maker intended.

Viewing trained tags is also a very useful way to figure out what to use as a trigger word, and in my experience, is actually more reliable than what's listed on CivitAI in some cases.  Other UIs like SwarmUI, and Auto1111 have the ability to view embedded metadata built in, but Comfy does not, and it's a real shame.

It would also be nice to be able to customize some basic metadata within SwarmUI's image drawer.  Basic things like a button for models to set the model's displayed name, as well as the ablity to browse for a file to upload, or pick an image from the output folder, which can be used as a model preview.

**Comment by JaredTherriault** [comment]
Ok, so I just pushed an update that does something like this.

In the event that a lora doesn't have a user-set friendly name or a civit info file, it will use the name from the embedded modelspec (or output name) in the safestensor's metadata.

Similarly, trained words/tags check user-set values first, then check for a civit info file, then if one isn't found just uses the tag frequency keys from the ST meta.

Finally, the metadata is added to the "tooltip" title property for each lora card, if the meta exists.

I do have plans for a form that will allow the user to save custom names, trained words, images, etc and it will be a good place to show meta and info files in a friendlier way than just as a tooltip. But this isn't an immediate concern. I'll keep this issue open as a reminder.

**Comment by Monty1836** [comment]
Thanks, I like the button to copy both lora name, and trained words at the same time, it seems like it'll be very useful once you can customize them.  I notice the first time you load models, it takes a lot longer now, but not after the first time.  I'm curious, are you caching them somewhere, or is python doing something automatically?  If it's simple to implement, caching them to the civit info files that the image drawer already reads would actually make them user editable in a fairly simple way, and I assume it would still be useful for the editing UI later.

On a somewhat unrelated note, thank you for the work you put into this extension.  It's probably my favorite extension for ComfyUI, and everything you do to keep making it better only make me like it even more.  I think it says a lot that some of your functionality has even been officially implemented into ComfyUI, and isn't needed here anymore.  You're doing excellent work!

**Comment by JaredTherriault** [comment]
Thanks monty! I implemented a rudimentary info editor (the pencil icon) for loras and embeddings. Give it a try, let me know if you see any issues or can think of anything else you'd like to add to the edit form.

**Comment by Monty1836** [comment]
I like it!  There are more options than I'd expected here, and I like all the overrides that are available.  Defaults for model, and clip strengths per-model in specific seem very useful, and make me wonder why it's not something I've seen before.

The button to open the folder containing the model also seems nice to have, but currently doesn't work for me.  It just opens the desktop instead, so I assume that path isn't resolving for some reason.

Taking longer to load is definitely worth having metadata accessible in my opinion, though rather than caching the all of the model's metadata, why not use it to fill the friendly name, and trained words in the user info just once if neither a user info or civit info file exist, then add a button to view the raw metadata from inside the user info editor?  That way you only need to check metadata once, and if the user wants to access it to customize their user info, you can load just the one model the user is looking at.

**Comment by JaredTherriault** [comment]
> The button to open the folder containing the model also seems nice to have, but currently doesn't work for me. It just opens the desktop instead, so I assume that path isn't resolving for some reason.

This can be a tricky fix. I'm developing on Arch Linux and it works, but I don't have access to mac or Windows to test directly. I added a logging event to this function. Can you update JNodes, try again, and send me the logs from Comfy?

Also, do you have this same issue when trying to open an image in File Manager from the feed or input/output contexts?

**Comment by Monty1836** [comment]
Nope, no problems opening images in file manager, it's something I use constantly, and it's always worked.  Double checked, and it still does.    Based on the log output, looks like the issue is that you're using the wrong slashes!  I assume you've already got some way of compensating for this in the images, so hopefully you can just copy that code.

JNodes: opening file manager on Windows at path: RemovedBeforePosting\ComfyUI_windows_portable\ComfyUI\models\loras\SDXL/pk_trainer_xl_v1_(pixelart).safetensors

**Comment by JaredTherriault** [comment]
Of course. Curse you Windows! *shakes fist*

Path should be normalized now, go ahead and update.

**Comment by Monty1836** [comment]
Just tested, looks like it's working now.  Thanks!  That's just how programming is really, you inevitably forget about something small like that.
